src/api/java/com/minecolonies/api/colony/IColonyTagCapability.java

```diff
@@ -178,33 +207,53 @@ public void addBuildingClaim(final int colonyId, final BlockPos pos, final Level
         @Override
         public void removeBuildingClaim(final int colonyId, final BlockPos pos, final LevelChunk chunk)
         {
-            if (claimingBuildings.containsKey(colonyId))
+            if (!claimingBuildings.containsKey(colonyId))
             {
-                final Set<BlockPos> buildings = claimingBuildings.get(colonyId);
-                buildings.remove(pos);
+                return;
+            }
 
-                if (buildings.isEmpty())
-                {
-                    claimingBuildings.remove(colonyId);
-                }
+            chunk.setUnsaved(true);
+            final Set<BlockPos> buildings = claimingBuildings.get(colonyId);
+            buildings.remove(pos);
+
+            if (buildings.isEmpty())
+            {
+                claimingBuildings.remove(colonyId);
 
                 if (owningColony == colonyId)
                 {
                     if (claimingBuildings.isEmpty())
                     {
-                        reset(chunk);
-                    }
-                    else if (claimingBuildings.size() == 1)
-                    {
-                        setOwningColony(claimingBuildings.keySet().iterator().next(), chunk);
+                        if (!colonies.contains(owningColony))
+                        {
+                            if (colonies.isEmpty())
+                            {
+                                owningColony = NO_COLONY_ID;
+                            }
+                            else
+                            {
+                                owningColony = colonies.iterator().next();
+                            }
+                        }
                     }
                     else
                     {
-                        setOwningColony(claimingBuildings.keySet().toArray(new Integer[0])[new Random().nextInt(claimingBuildings.size())], chunk);
+                        for (final Map.Entry<Integer, Set<BlockPos>> colonyEntry : claimingBuildings.entrySet())
+                        {
+                            final IColony colony = IColonyManager.getInstance().getColonyByDimension(colonyEntry.getKey(), chunk.getLevel().dimension());
```

Comment:
maybe we do a) The colony null check before the loop, and b) We remove the buildings when they're null?